File: backend/services/fraud_sentinel.py
from typing import List, Dict, Optional
from datetime import datetime, timedelta

class VelocityRule:
    def __init__(self, rule_id: str, description: str, window_minutes: int, max_txn_count: int, max_amount_paise: int, segments: List[str], action: str):
        self.rule_id = rule_id
        self.description = description
        self.window_minutes = window_minutes
        self.max_txn_count = max_txn_count
        self.max_amount_paise = max_amount_paise
        self.segments = segments
        self.action = action # 'BLOCK' | 'STEP_UP_AUTH' | 'ALERT_RM'

# Velocity Rules as specified in blueprint
VELOCITY_RULES = [
    VelocityRule("VR001", "High-frequency small transactions (structuring pattern)", 60, 10, 50000000, ["ALL"], "BLOCK"),
    VelocityRule("VR002", "Multiple large transfers in 24h", 1440, 3, 500000000, ["ALL"], "STEP_UP_AUTH"),
    VelocityRule("VR003", "Senior: any unusual large outflow", 1440, 2, 50000000, ["SENIOR"], "ALERT_RM"),
    VelocityRule("VR004", "Senior: new beneficiary + large transfer", 60, 1, 10000000, ["SENIOR"], "BLOCK"),
    VelocityRule("VR005", "Midnight hour transaction (11PM–5AM) for Senior", 60, 5, 20000000, ["SENIOR"], "STEP_UP_AUTH"),
]

def check_velocity_rules(
    user_id: str, segment: str, recent_txns: List[Dict],
    new_txn_amount_paise: int, new_txn_timestamp: datetime,
    is_new_beneficiary: bool = False
) -> List[Dict]:
    """
    Evaluates velocity rules based on user segment and recent transactions.
    """
    signals = []
    
    applicable_rules = [
        r for r in VELOCITY_RULES 
        if "ALL" in r.segments or segment.upper() in r.segments
    ]
    
    for rule in applicable_rules:
        # Check rule condition filters
        if rule.rule_id == "VR004" and not is_new_beneficiary:
            continue
        if rule.rule_id == "VR005":
            hour = new_txn_timestamp.hour
            if hour >= 5 and hour < 23:
                continue
                
        window_start = new_txn_timestamp - timedelta(minutes=rule.window_minutes)
        window_txns = []
        for t in recent_txns:
            # Parse timestamp if string
            ts = t.get("txn_timestamp")
            if isinstance(ts, str):
                try:
                    ts = datetime.fromisoformat(ts.replace("Z", "+00:00"))
                except ValueError:
                    ts = new_txn_timestamp # Fallback
            if ts and ts >= window_start:
                window_txns.append(t)
                
        window_count = len(window_txns)
        window_amount = sum(t.get("amount_paise", 0) for t in window_txns) + new_txn_amount_paise
        
        if window_count >= rule.max_txn_count or window_amount >= rule.max_amount_paise:
            # Calculate normalization score
            risk_score = min(
                (window_count / rule.max_txn_count + window_amount / rule.max_amount_paise) / 2.0,
                1.0
            )
            signals.append({
                "rule_id": rule.rule_id,
                "signal_type": "VELOCITY",
                "severity": "HIGH" if risk_score > 0.7 else "MEDIUM",
                "risk_score": round(risk_score, 4),
                "description": rule.description,
                "action": rule.action,
                "auto_blocked": rule.action == "BLOCK"
            })
            
    return signals

def verify_device_fingerprint(
    user_id: str, segment: str, incoming_hash: str, trusted_hashes: List[str]
) -> Optional[Dict]:
    """
    Verifies if device fingerprint is recognized. For seniors, unrecognized device prompts block.
    """
    if not trusted_hashes or incoming_hash in trusted_hashes:
        return None
        
    severity = "HIGH" if segment.upper() == "SENIOR" else "MEDIUM"
    action = "BLOCK" if segment.upper() == "SENIOR" else "STEP_UP_AUTH"
    
    return {
        "signal_type": "DEVICE",
        "severity": severity,
        "risk_score": 0.85 if segment.upper() == "SENIOR" else 0.60,
        "description": "Unrecognized device fingerprint detected.",
        "action": action,
        "auto_blocked": action == "BLOCK"
    }

def score_biometric_anomaly(
    user_id: str, segment: str,
    avg_dwell_ms: float, baseline_dwell_ms: float,
    baseline_variance: float
) -> Optional[Dict]:
    """
    Checks if session biometric keystroke dynamics deviate significantly from baseline.
    """
    if baseline_variance <= 0:
        baseline_variance = 1.0
    deviation = abs(avg_dwell_ms - baseline_dwell_ms)
    z_score = deviation / (baseline_variance ** 0.5)
    
    if z_score > 3.0:
        return {
            "signal_type": "BIOMETRIC",
            "severity": "HIGH" if segment.upper() == "SENIOR" else "MEDIUM",
            "risk_score": min(z_score / 5.0, 1.0),
            "description": f"Behavioral biometrics z-score deviation too high (z={z_score:.2f}). Possible account takeover.",
            "action": "STEP_UP_AUTH",
            "auto_blocked": False
        }
    return None

def check_anomaly(amount: float, velocity_24h: int, location: str):
    # Backward compatibility with existing route
    if velocity_24h > 5 or amount > 50000 or location not in ["India", "IN"]:
        return True
    return False

File: frontend/js/fraud-sentinel.js
/**
 * Avatar Base — Fraud Sentinel (Frontend)
 * Real-time velocity rules, device fingerprinting, biometric anomaly scoring
 */

// ── Velocity Rules ────────────────────────────────────────────────────
const VELOCITY_RULES = [
  {
    id: 'VR001', desc: 'High-frequency small transactions (structuring pattern)',
    windowMin: 60, maxCount: 10, maxAmount: 500000, segments: ['ALL'], action: 'BLOCK'
  },
  {
    id: 'VR002', desc: 'Multiple large transfers in 24h',
    windowMin: 1440, maxCount: 3, maxAmount: 5000000, segments: ['ALL'], action: 'STEP_UP_AUTH'
  },
  {
    id: 'VR003', desc: 'Senior: any unusual large outflow',
    windowMin: 1440, maxCount: 2, maxAmount: 500000, segments: ['SENIOR'], action: 'ALERT_RM'
  },
  {
    id: 'VR004', desc: 'Senior: new beneficiary + large transfer',
    windowMin: 60, maxCount: 1, maxAmount: 100000, segments: ['SENIOR'], action: 'BLOCK'
  },
  {
    id: 'VR005', desc: 'Midnight hour transaction (11PM–5AM) for Senior',
    windowMin: 60, maxCount: 5, maxAmount: 200000, segments: ['SENIOR'], action: 'STEP_UP_AUTH'
  },
];

// ── Velocity Check ────────────────────────────────────────────────────
export function checkVelocityRules({ segment, recentTxns, newTxnAmount, newTxnTimestamp, isNewBeneficiary = false }) {
  const signals = [];
  const now = new Date(newTxnTimestamp);

  const applicable = VELOCITY_RULES.filter(r => r.segments.includes('ALL') || r.segments.includes(segment));

  for (const rule of applicable) {
    if (rule.id === 'VR004' && !isNewBeneficiary) continue;
    if (rule.id === 'VR005') {
      const hour = now.getHours();
      if (hour >= 5 && hour < 23) continue;
    }

    const windowStart = new Date(now.getTime() - rule.windowMin * 60 * 1000);
    const windowTxns = recentTxns.filter(t => new Date(t.timestamp) >= windowStart);
    const windowCount = windowTxns.length;
    const windowAmount = windowTxns.reduce((s, t) => s + (t.amount || 0), 0) + newTxnAmount;

    if (windowCount >= rule.maxCount || windowAmount >= rule.maxAmount) {
      const riskScore = Math.min(
        (windowCount / rule.maxCount + windowAmount / rule.maxAmount) / 2, 1.0
      );
      signals.push({
        ruleId: rule.id,
        signalType: 'VELOCITY',
        severity: riskScore > 0.7 ? 'HIGH' : 'MEDIUM',
        riskScore: parseFloat(riskScore.toFixed(4)),
        description: rule.desc,
        action: rule.action,
        autoBlocked: rule.action === 'BLOCK',
      });
    }
  }
  return signals;
}

// ── Device Fingerprinting (Browser-side) ──────────────────────────────
export async function captureDeviceFingerprint() {
  const fp = {
    userAgent:   navigator.userAgent,
    platform:    navigator.platform || 'unknown',
    language:    navigator.language,
    timezone:    Intl.DateTimeFormat().resolvedOptions().timeZone,
    screen:      `${screen.width}x${screen.height}x${screen.colorDepth}`,
    concurrency: navigator.hardwareConcurrency || 0,
    cookiesEnabled: navigator.cookieEnabled,
    doNotTrack:  navigator.doNotTrack || 'unknown',
    touchPoints: navigator.maxTouchPoints || 0,
  };

  const fpString = JSON.stringify(fp);
  // SHA-256 via Web Crypto API
  const encoded = new TextEncoder().encode(fpString);
  const hashBuffer = await crypto.subtle.digest('SHA-256', encoded);
  const hashArray = Array.from(new Uint8Array(hashBuffer));
  const fingerprint = hashArray.map(b => b.toString(16).padStart(2, '0')).join('');

  return { fingerprint, metadata: fp };
}

// ── Behavioral Biometrics ─────────────────────────────────────────────
export class BiometricCollector {
  constructor() {
    this.keyEvents = [];
    this.swipeEvents = [];
    this.tapEvents = [];
    this._listening = false;
  }

  start() {
    if (this._listening) return;
    this._listening = true;
    this._onKeydown = (e) => this.keyEvents.push({ key: e.code, time: Date.now(), type: 'down' });
    this._onKeyup   = (e) => this.keyEvents.push({ key: e.code, time: Date.now(), type: 'up' });
    this._onTouch   = (e) => this.tapEvents.push({ x: e.touches[0]?.clientX, y: e.touches[0]?.clientY, time: Date.now() });
    document.addEventListener('keydown', this._onKeydown);
    document.addEventListener('keyup', this._onKeyup);
    document.addEventListener('touchstart', this._onTouch, { passive: true });
  }

  stop() {
    document.removeEventListener('keydown', this._onKeydown);
    document.removeEventListener('keyup', this._onKeyup);
    document.removeEventListener('touchstart', this._onTouch);
    this._listening = false;
  }

  getMetrics() {
    const pairs = [];
    for (let i = 0; i < this.keyEvents.length - 1; i++) {
      const cur = this.keyEvents[i];
      const nxt = this.keyEvents[i + 1];
      if (cur.type === 'down' && nxt.type === 'up' && cur.key === nxt.key) {
        pairs.push(nxt.time - cur.time); // dwell time
      }
    }
    const avgDwell = pairs.length ? pairs.reduce((s, v) => s + v, 0) / pairs.length : 120;
    const variance = pairs.length > 1
      ? pairs.reduce((s, v) => s + (v - avgDwell) ** 2, 0) / pairs.length : 50;

    return {
      avgKeystrokeDwellMs: parseFloat(avgDwell.toFixed(2)),
      keystrokeVariance: parseFloat(variance.toFixed(2)),
      sampleCount: pairs.length,
    };
  }

  /**
   * Compute anomaly z-score vs. historical baseline
   * Returns { isAnomalous, zScore }
   */
  scoreAnomaly(historical = { avgDwell: 120, variance: 50 }) {
    const metrics = this.getMetrics();
    if (metrics.sampleCount < 5) return { isAnomalous: false, zScore: 0 };
    const deviation = Math.abs(metrics.avgKeystrokeDwellMs - historical.avgDwell);
    const zScore = deviation / Math.max(Math.sqrt(historical.variance), 1);
    return {
      isAnomalous: zScore > 3.0,
      zScore: parseFloat(zScore.toFixed(2)),
      metrics,
    };
  }
}

// ── Anomaly Risk Score (Isolation Forest approximation) ───────────────
export function computeIsolationScore(features) {
  /**
   * Browser-side lightweight anomaly scorer.
   * Normalizes 6 behavioral features and returns a risk score 0-1.
   * A full Isolation Forest runs on the backend; this provides fast UX-level feedback.
   */
  const {
    txnAmount = 0,          avgTxnAmount = 5000,
    hourOfDay = 12,         typicalHour = 14,
    newMerchant = false,    merchantFamiliarityScore = 1.0,
    velocityZ = 0,
    deviceTrustScore = 1.0,
    biometricZ = 0,
  } = features;

  const amountZ   = Math.min(Math.abs(txnAmount - avgTxnAmount) / Math.max(avgTxnAmount, 1), 3) / 3;
  const hourZ     = Math.min(Math.abs(hourOfDay - typicalHour) / 12, 1);
  const merchantZ = newMerchant ? (1 - merchantFamiliarityScore) : 0;
  const velZ      = Math.min(velocityZ / 3, 1);
  const devZ      = 1 - deviceTrustScore;
  const bioZ      = Math.min(biometricZ / 5, 1);

  // Weighted composite
  const score = (amountZ * 0.20 + hourZ * 0.15 + merchantZ * 0.20 +
                 velZ * 0.20 + devZ * 0.15 + bioZ * 0.10);

  const severity = score > 0.7 ? 'CRITICAL' : score > 0.5 ? 'HIGH' :
                   score > 0.3 ? 'MEDIUM' : 'LOW';

  return { score: parseFloat(score.toFixed(4)), severity, breakdown: { amountZ, hourZ, merchantZ, velZ, devZ, bioZ } };
}

// ── Fraud Alert Renderer ──────────────────────────────────────────────
export function renderFraudBanner(container, signals, segment) {
  if (!signals || signals.length === 0) {
    container.style.display = 'none';
    return;
  }

  const top = signals.sort((a, b) => b.riskScore - a.riskScore)[0];
  container.style.display = 'flex';
  container.innerHTML = `
    <div class="fraud-banner-icon">🚨</div>
    <div class="fraud-banner-content">
      <div class="fraud-banner-title">
        ${top.autoBlocked ? 'Transaction Blocked for Your Safety' : 'Unusual Activity Detected'}
      </div>
      <div class="fraud-banner-sub">
        ${top.description} · Risk score: ${(top.riskScore * 100).toFixed(0)}% · 
        Action: ${top.action.replace(/_/g, ' ')}
      </div>
    </div>
    <button class="btn btn-danger btn-sm" onclick="emergencyBlock()">
      🛑 Block Account
    </button>
    ${segment === 'SENIOR' ? '<button class="btn btn-secondary btn-sm ml-2" onclick="callRM()">📞 Call RM</button>' : ''}
  `;
}

function emergencyBlock() {
  alert('🚨 Your account has been temporarily blocked for safety. Please call the bank helpline: 1800-XXX-XXXX');
}

function callRM() {
  alert('📞 Connecting you to your Relationship Manager. RM will call your registered number within 2 minutes.');
}
